scvi/external/sysvi/_trainingplans.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import annotations
+
+from inspect import getfullargspec
+from typing import Literal
+
+import torch
+
+from scvi.module.base import BaseModuleClass, LossOutput
+from scvi.train import TrainingPlan
+
+# TODO could make new metric class to not be called elbo metric as used for other metrics as well
+from scvi.train._metrics import ElboMetric
+
+
+class WeightScaling:
+    """Linearly scale loss weights between start and end weight accordingly to the current training stage
+
+    Parameters
+    ----------
+    weight_start
+        Starting weight value
+    weight_end
+        End weight vlue
+    point_start
+        Training point to start scaling - before weight is weight_start
+        Since the epochs are counted after they are run,
+        the start point must be set to 0 to represent 1st epoch
+    point_end
+        Training point to end scaling - after weight is weight_end
+        Since the epochs are counted after they are run,
+        the start point must be set to n-1 to represent the last epoch
+    update_on
+        Define training progression based on epochs or steps
+
+    """
+
+    def __init__(
+        self,
+        weight_start: float,
+        weight_end: float,
+        point_start: int,
+        point_end: int,
+        update_on: Literal["epoch", "step"] = "step",
+    ):
+        self.weight_start = weight_start
+        self.weight_end = weight_end
+        self.point_start = point_start
+        self.point_end = point_end
+        if update_on not in ["step", "epoch"]:
+            raise ValueError("update_on not recognized")
+        self.update_on = update_on
+
+        weight_diff = self.weight_end - self.weight_start
+        n_points = self.point_end - self.point_start
+        self.slope = weight_diff / n_points
+
+        if (
+            self.weight(epoch=self.point_start, step=self.point_start) < 0
+            or self.weight(epoch=self.point_end, step=self.point_end) < 0
+        ):
+            raise ValueError("Specified weight scaling would lead to negative weights")
+
+    def weight(
+        self,
+        epoch: int,
+        step: int,
+    ) -> float:
+        """
+        Computes the weight for the current step/epoch depending on which update type was set in init
+
+        Parameters
+        ----------
+        epoch
+            Current epoch.
+        step
+            Current step.
+        """
+        if self.update_on == "epoch":
+            point = epoch
+        elif self.update_on == "step":
+            point = step
+        else:
+            # This is ensured not to happen by above init check
+            raise ValueError("self.update_on not recognised")
+
+        if point < self.point_start:
+            return self.weight_start
+        elif point > self.point_end:
+            return self.weight_end
+        else:
+            return self.slope * (point - self.point_start) + self.weight_start
+
+
+class TrainingPlanCustom(TrainingPlan):
```

Comment:
Increase inheritance. Some of the new functions are duplicates of TrainingPlan functions.